File: haskell/src/Solutions/Euler022.hs
module Solutions.Euler022 where
import Utils.Input(getInput)
import Utils.List(splitOn)
import Data.List(sort)
import Data.Char(ord)

wordValue :: String -> Int
wordValue = sum . map (\c -> ord c - ord 'A' + 1)

pairValue :: Int -> String -> Int
pairValue x w = wordValue w * x

euler022 :: IO String
euler022 = show . sum . zipWith pairValue [1..] . sort . map (tail . init) . splitOn ',' <$> getInput 22

File: rust/src/solutions/euler022.rs
use crate::utils::input::get_input;

fn word_value(word: &str) -> i64 {
    word.chars().map(|c| c as i64 - 'A' as i64 + 1).sum()
}

pub fn euler022() -> String {
    let mut array: Vec<String> = get_input(22)
        .split(',')
        .map(|l| l.trim_matches('"').to_string())
        .collect();
    array.sort();
    array
        .iter()
        .map(|w| word_value(&w))
        .zip(1..)
        .map(|(v, i)| v * i)
        .sum::<i64>()
        .to_string()
}
